src/resolv/lookup/srv.rs

```diff
@@ -0,0 +1,274 @@
+//! Looking up SRV records.
+
+use futures::{Async, Future, Poll};
+use std::net::IpAddr;
+use ::bits::{DName, DNameBuf, DNameSlice, MessageBuf, ParsedDName, ParseResult, ComposeError};
+use ::iana::{Rtype, Class};
+use ::rdata::{Srv, A, Aaaa};
+use super::host::{lookup_host, LookupHost, FoundHosts};
+use super::super::{Query, Resolver};
+use super::super::error::{Error, Result};
+use super::search::SearchIter;
+use rand;
+use rand::distributions::{IndependentSample, Range};
+
+
+//------------ lookup_records ------------------------------------------------
+
+/// Creates a future that looks up SRV records.
+///
+/// TODO: documentation
+pub fn lookup_srv<N>(resolver: Resolver, txt_service: N, name: N, fallback_port: u16)
+                     -> Result<LookupSRVState>
+                  where N: AsRef<DNameSlice> {
+    let name : &DNameSlice = name.as_ref();
+    let txt_service : &DNameSlice = txt_service.as_ref();
+    let full_name = match txt_service.join(&name) {
+        Ok(full_name) => full_name,
+        Err(_) => return Err(Error::Question(ComposeError::LongName))
+    };
+    let mut search = SearchIter::new(resolver.clone(), &full_name);
+    let search_name = search.as_mut().map(|s| s.next().unwrap());
+    let query_name = match search_name {
+        Some(ref name) => name,
+        None => &full_name
+    };
+    let query = resolver.clone().query((query_name, Rtype::Srv, Class::In));
+    Ok(LookupSRVState::LookupSRV(LookupSRV {
+        resolver: resolver.clone(),
+        host: name.to_cow().into_owned(),
+        fallback_port: fallback_port,
+        query: query,
+        search: search
+    }))
+}
+
+
+//------------ LookupRecords -------------------------------------------------
+
+/// The future returned by [`lookup_srv()`].
+///
+/// [`lookup_srv()`]: fn.lookup_srv.html
+pub struct LookupSRV {
+    /// The resolver to run queries on.
+    resolver: Resolver,
+
+    /// Bare host to be queried, used for fallback if no SRV records are found
+    host: DNameBuf,
+
+    /// Fallback port, used if no SRV records are found
+    fallback_port: u16,
+
+    /// The current SRV query
+    query: Query,
+
+    /// An optional search list iterator for searching a name.
+    search: Option<SearchIter>,
+}
+
+pub enum LookupSRVState {
+    LookupSRV(LookupSRV),
+    FallbackHost(LookupHost, u16),
+}
+
+impl Future for LookupSRVState {
+    type Item = LookupResult;
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        *self = match *self {
+            LookupSRVState::LookupSRV(ref mut a) =>
+                match a.poll() {
+                    Ok(item) => return Ok(item),
+                    Err(_) => LookupSRVState::FallbackHost(lookup_host(a.resolver.clone(), &a.host), a.fallback_port)
+                },
+            LookupSRVState::FallbackHost(ref mut a, fallback_port) =>
+                match a.poll() {
```

Comment:
In this case, we don't want to early return on error, as we want to use the fallback instead.